warehouse/accounts/views.py

```diff
@@ -232,7 +228,99 @@ def two_factor(request, _form_class=TwoFactorForm):
 
             return resp
 
-    return {"form": form}
+    return two_factor_state
+
+
+@view_config(
+    uses_session=True,
+    request_method="GET",
+    route_name="accounts.webauthn-authenticate.options",
+    renderer="json",
+)
+def webauthn_authentication_options(request):
+    if request.authenticated_userid is not None:
+        return {"fail": {"errors": ["Already authenticated"]}}
+
+    try:
+        two_factor_data = _get_two_factor_data(request)
+    except TokenException:
+        request.session.flash("Invalid or expired two factor login.", queue="error")
+        return {"fail": {"errors": ["Invalid two factor token"]}}
+
+    userid = two_factor_data.get("userid")
+    user_service = request.find_service(IUserService, context=None)
+    return user_service.get_webauthn_assertion_options(
+        userid,
+        challenge=request.session.get_webauthn_challenge(),
+        icon_url=request.registry.settings.get(
+            "warehouse.domain", request.domain
+        ),
+        rp_id=request.domain,
+    )
+
+
+@view_config(
+    require_csrf=True,
+    require_methods=False,
+    uses_session=True,
+    request_method="POST",
+    request_param=WebAuthnAuthenticationForm.__params__,
+    route_name="accounts.webauthn-authenticate.validate",
+    renderer="json"
+)
+def webauthn_authentication_validate(request):
```

Comment:
Pinging @di @dstufft @ewdurbin for feedback here:

The WebAuthn authentication flow requires a separate form and renderer, since we `POST` from client-side JS. However, this leads to some confusing naming above: `login` calls `two_factor`, which renders its template with the TOTP form and/or the WebAuthn button depending on what the user has.

The user then either `POST`s back to `two_factor` (for TOTP), or `POST`s to `webauthn_authentication_validate` (for WebAuthn). This is a little confusing, since `two_factor` is **only** the TOTP `POST`-back.

Some ideas:
* Split `two_factor` into two methods: one `GET` that renders the TOTP form + WebAuthn button, and one `POST` (named `totp_authentication_validate` or similar) that validates the TOTP `POST`-back. This makes it more obvious as to what is doing what, but it also means that we have to juggle the TOTP authentication form in two methods instead of one.
* Rename `two_factor` to `two_factor_and_totp_authenticate` or something similar. Not very pleasant to read.
* Document the current behavior in comments. Involves the fewest code changes, but retains the slightly confusing naming.

I'm not completely happy with any of these options, so I'd appreciate your insight.

Reply:
I don't have a preference, though the second option seems most reasonable to me. Even if it isn't pleasant, it is explicit!

Reply:
Sounds good! I'll make this change while adding support for multiple WebAuthn credentials.